libsolidity/ast/TypeProvider.h

```diff
@@ -0,0 +1,235 @@
+ *
+ * It is not recommended to explicitly instanciate types unless you really know what and why
+ * you are doing it.
+ */
+class TypeProvider
+{
+public:
+	TypeProvider();
+	TypeProvider(TypeProvider&&) = default;
+	TypeProvider(TypeProvider const&) = delete;
+	TypeProvider& operator=(TypeProvider&&) = default;
+	TypeProvider& operator=(TypeProvider const&) = delete;
+	~TypeProvider() = default;
+
+	/**
+	 * Global TypeProvider instance.
+	 *
+	 * This is only available initially and is going to be factored out in future PRs (TODO),
+	 * so we keep the initial PR lean and clean.
+	 */
+	static TypeProvider& get()
+	{
+		static TypeProvider _provider;
+		return _provider;
+	}
+
+	Type const* fromElementaryTypeName(ElementaryTypeNameToken const& _type);
+
+	BoolType const* boolType() const noexcept { return &m_boolType; }
+
+	FixedBytesType const* byteType() { return fixedBytesType(1); }
+	FixedBytesType const* fixedBytesType(unsigned m) { return &m_bytesM.at(m - 1); }
+
+	ArrayType const* bytesType() const noexcept { return &m_bytesType; }
+	ArrayType const* bytesMemoryType() const noexcept { return &m_bytesMemoryType; }
+	ArrayType const* stringType() const noexcept { return &m_stringType; }
+	ArrayType const* stringMemoryType() const noexcept { return &m_stringMemoryType; }
+
+	/// Constructor for a byte array ("bytes") and string.
+	ArrayType const* arrayType(DataLocation _location, bool _isString = false);
+
+	/// Constructor for a dynamically sized array type ("type[]")
+	ArrayType const* arrayType(DataLocation _location, Type const* _baseType);
+
+	/// Constructor for a fixed-size array type ("type[20]")
+	ArrayType const* arrayType(DataLocation _location, Type const* _baseType, u256 const& _length);
+
+	AddressType const* payableAddressType() const noexcept { return &m_payableAddressType; }
+	AddressType const* addressType() const noexcept { return &m_addressType; }
+
+	IntegerType const* intType(unsigned m = 256) const
+	{
+		solAssert((m % 8) == 0, "");
+		return &m_intM.at(m / 8 - 1);
+	}
+
+	IntegerType const* uintType(unsigned m = 256) const
+	{
+		solAssert((m % 8) == 0, "");
+		return &m_uintM.at(m / 8 - 1);
+	}
+
+	IntegerType const* integerType(unsigned _bits = 256, IntegerType::Modifier _modifier = IntegerType::Modifier::Unsigned)
+	{
+		return _modifier == IntegerType::Modifier::Unsigned ? uintType(_bits) : intType(_bits);
+	}
+
+	FixedPointType const* fixedType() { return fixedType(128, 18); }
+	FixedPointType const* fixedType(unsigned m, unsigned n);
+
+	FixedPointType const* ufixedType() { return ufixedType(128, 18); }
+	FixedPointType const* ufixedType(unsigned m, unsigned n);
+
+	FixedPointType const* fixedPointType(unsigned m, unsigned n, FixedPointType::Modifier _modifier)
+	{
+		return _modifier == FixedPointType::Modifier::Unsigned ? ufixedType(m, n) : fixedType(m, n);
+	}
+
+	StringLiteralType const* stringLiteralType(std::string const& literal);
+
+	TupleType const* tupleType(std::vector<Type const*>&& members);
```

Comment:
Why not just pass by value?

Reply:
I agree with having this question. I first checked what cases we have on the caller side, and none of them re-used the passed parameter after the tupleType() call. Passing by value though would give us the freedom to allow both cases, I agree.